### ISSUE TITLE
Implement disc-scanning

### ISSUE THREAD

**Issue by WizLizard** [Asnivor/MedLaunch#69]
Even if i set my psx game folder path in the settings tab and then switch to the games tab and scan for my roms, disc based image files are not detected and added. Why is this? The only way I'm able to do it is manually. Right click psx under game tab then add game manually finally selecting a .cue or .m3u file.

**Comment by Asnivor** [comment]
Sorry, auto-disc scanning is not yet supported (although it is planned for soon). Its manual only until I get it implemented. Shouldn't be too long

**Comment by WizLizard** [comment]
Ok could you make it so that .cue, .bin and .m3u files are detected in sub
folders when scanning? Because i like to separate my ps1 games with sub
folders

**Comment by Asnivor** [comment]
That was what I was planning. I actually do the same thing but was unsure what other people tend to do.

**Comment by WizLizard** [comment]
Awesome thank you very much, glad I'm not the only one lol. Keep up the
good work. This is by far the best frontend for mednafen hands down and
I've tried them all believe me. If you get this feature added along with
full fledged controller mapping it will be even better!

**Comment by Asnivor** [comment]
The controller mapping will be the biggy :)

I'll get there eventually though. Thanks!

**Comment by WizLizard** [comment]
Good luck i look forward to it. Sorry one more thing, just wanted to know
if there is a way to individually hide game systems visible on thre games
tab and configs tab?

**Comment by Asnivor** [comment]
There is not, but a show/hide systems selector should be trivial enough to implement...

**Comment by WizLizard** [comment]
Cool sounds good. The frontend mednaffe has something similar. It's been my
go to frontend but I think thats gonna change with this frontend. The
feature comes in handy if you don't intend to use every core plus it cleans
things up a bit. Thanks again I'm glad to have stumbled upon this frontend.

**Comment by WizLizard** [comment]
Any chance of this making it to the next release?

**Comment by Asnivor** [comment]
I'm hoping so. This and controller configuration is my aim for next release.

**Comment by retro-wertz** [comment]
psx, pcfx, pce-cd, supercd have identifications like headers or keyword inside the data tracks, so scanning folders would with assorted stuff should allow auto-assigning the files into appropriate folders. im just not sure how hard this would be to code.

the fastest way for end-user is to organize the cue-sheets in approriate folders/systems before hard and scan individually by system. this would be faster to code as it only needs to find the .cue, .ccd or .toc and not bother messing with the data tracks. 

IMO

**Comment by Asnivor** [comment]
> psx, pcfx, pce-cd, supercd have identifications like headers or keyword inside the data tracks, so scanning folders would with assorted stuff should allow auto-assigning the files into appropriate folders. im just not sure how hard this would be to code.

Its entirely possible, but will require quite a bit of faffing.

> the fastest way for end-user is to organize the cue-sheets in approriate folders/systems before hard and scan individually by system. this would be faster to code as it only needs to find the .cue, .ccd or .toc and not bother messing with the data tracks.

That is pretty much what I am going to be aiming for. User selects say a PSX folder in MedLaunch. Then when PSX scan is started it looks for sub-folders within this folder. Each sub folder should contain 1 game only. The scanner will look for .m3u files first, if one is found then this game will be added to the library (and the m3u interogated to ascertain any cue/ccd or toc files specified. At this point I can derive the actual image files and look at the data track if neccesary (eg, to get the serial number for PSX games). If m3u is not discovered then it will look for a single cue/ccd/toc and add this. If there is more than one then I can do some string comparison (maybe looking for similar names with variations at the end (1, 2, 3 - CD1, CD2, DISC 1 etc..) and auto-generate an m3u file.

Once something like this is in place I can possibly think about also catering for (crazy) people who just have one folder (without any subfolders) that contains all their cue/images files for the same system.

Anyone who keeps cue/images for different systems in the same folder will probably not be catered for :)

**Comment by WizLizard** [comment]
Really looking forward to this feature will make my life and others a lot easier, I have a lot of ps1 games lol.

**Comment by Asnivor** [comment]
This is now implemented and working (and will be in the next release), The following must be true though:

- Each game (collection of disk image and cue/ccd/toc files) MUST reside in its own sub-folder - the parent folder being the one you set as the system path in MedLaunch settings
- Any game files in the root of the system folder will be ignored
- Multiple games within the same sub-folder will likely cause problems (any random files that are not .m3u, .cue, .ccd or .toc will just be ignored anyway)
- The sub-folders should be named descriptively (at the moment the sub-folder name is used as the game name in the MedLaunch games library in the case of M3U playlist files)

A rough approximation of the scanning logic is as follows:

1. All sub-folders under the system folder are enumerated
2. Each of these is scanned for the presence of .m3u, .cue, .ccd and .toc files
3. If an existing .m3u file is found this is returned and imported/updated in the games library (processing for this subfolder is then terminated)
4. Next, the scanner searches for the presence of a single .cue/.ccd/.toc file (in that order) in the subdirectory where no other cuesheet files of other types are present (this handles one-disc games) and the game is returned and imported/updated in the games library (processing for this subfolder is then terminated)
5. Now, if the routine has gotten this far it means there are multiple cuesheet files (of potentially varying types) present in the subfolder. These are all scanned for the presence of numerals 1-6 in the filename and put into 6 separate lists
6. These lists are all looped though and the cuefiles interogated to make sure they have a valid image file referenced on disk (bin, img, iso etc). If they do, the first valid cuefile is taken (in the order .cue, .ccd, .toc)
7. We now should have a list of working cuefiles for this one game - this is ordered alphabetically - an .m3u playlist file generated and then imported/updated in the games library (processing for this subfolder is then terminated)

At the moment, this appears to be working fairly well in the testing I have done. I might in the future look at supporting multiple games in the same directory/subdirectory, but at the moment this is a little too much work...

**Comment by WizLizard** [comment]
Oh man thanks for your hard work! I can't wait to test it out!

**Comment by WizLizard** [comment]
So I tested the disc scanning and it detected all of my games no problem! However, scraping no longer seems to be working correctly. For my scraping options I only leave game manuals and box art checked. It missed about 4 games completely and for Driver the wrong game art and data was downloaded. It downloaded Pajama Sam Boxart and info. None of my game manuals were downloaded. All it says under game documents for each game is "request.php."

**Comment by Asnivor** [comment]
Strange. I'll take a look at the small changes I made to the scraping system.

 Was the auto-scraping working correctly before?

**Comment by Asnivor** [comment]
so the request.php manual thing. This appears to be because replacementdocs.com have a daily download quota (and some of the manuals are scraped from there). If you rescrape this single game a day after it will probably work. Not ideal though. Short of hosting them myself (which is not going to happen) I'm not sure of a solution at the moment.

Games that are missed completely in the autoscrape - this could be a bug somewhere, or it could just be that the game name isnt being auto-matched to an entry in the json file that it uses for lookups - https://github.com/Asnivor/MedLaunch/blob/master/MedLaunch/Data/System/TheGamesDB.json
If you select these missing games and click the 'identify and scrape game' button in the right panel, does a match appear in the list you can select?
If you can give me a list of these games that arent auto-scraping I can take a look to see if there are missing in the json file.

Re: wrong game data being downloaded - can you give me the exact filename of your Driver cue/ccd/toc file?

**Comment by WizLizard** [comment]
I'm going to check all the games that didn't scrape later tonight after
work and I'll get back to you.

**Comment by WizLizard** [comment]
The games not auto scraping correctly are Gran Turismo, Resident Evil Directors Cut,
Tomb Raider III The Adventures of Lara Croft and Driver You Are The Wheelman. The exact name of my driver .cue is Driver - You Are the Wheelman. There is a match for it that works when i do it manually though, same for the others.

**Comment by Asnivor** [comment]
So the way the auto-scraping logic works is kind of tricky. The method loads the previously mentioned json file into memory (https://github.com/Asnivor/MedLaunch/blob/master/MedLaunch/Data/System/TheGamesDB.json). Then it splits the game name into seperate words and passes this to a kind of fuzzy-search method that returns hits from the json file based on number of word hits (and an associated 'accuracy' score). Up to this point the manual 'pick game' method and auto-scrape methods use the same code.

Now, the auto-scrape function continues and attempts to find a *distinct* match from the returned results (using a horrible bit of convoluted procedural code - but thats another story). If there are no results matched, *or* if multiple are matched and no distinct has been found - then no game is auto-scraped.

Currently, this part of the code is iirc ignoring numbers (1, 2, 3 etc). It is also not converting roman numerals to numbers (I, II, III etc). I suspect that the reason some of your games will not auto-scrape (and nothing is found) is because *more* than one result is being found and the code cannot choose a distinct one from the results.

I'll work on this, but I am actually pretty close to having separate dat files (for psx and saturn) that includes game serial numbers and game titles. I can (in most cases) obtain the game serial number by interogating the disc image files (iso, bin, or img), so once this is in place the auto-scrape lookup can be modified to hopefully match up with my gamesdb json more accurately.

Re: driver being scraped as pajama sam - i really have no clue at the moment :/ I'm hoping the explanation will present itself at some point when I least expect it....
